File: lib/hand.js
'use strict'

let _ = require('lodash')
let d = require('debug')('holdem:hand')

var Hand = class hand {
  constructor(cards) {
    this.cards = cards
    this.hash = {} 
    this.suits = ''
    this.value = [] 
    this.rank = 0
    this.rankName = 0
    this.rankBit = 0

    this._reduce()
  }

  _toValue (cards) {
    cards = _.toString(cards)
    return cards.replace(/[CDHS]/g, '').split(',')
  }
  
  _faceValue (face) {
    var value = (face === 'T' ? 10 : face === 'J' ? 11 : face === 'Q' ? 12 : face === 'K' ? 13 : face === 'A' ? 14 : parseInt(face))
    return value
  }

  _reduce() {
    //extract suits
    this.suits = this.cards.toString().replace(/[^CDHS]/g, '')

    var hash = {}, i
    var value = this._toValue(this.cards).map((v) => {
      return this._faceValue(v)
    })

    for(i = 0; i < 5; i++) {
      hash[value[i]] = (hash[value[i]] >= 1) ? hash[value[i]] + 1 : 1
    }
    value = value.sort((a,b) => {
      return (hash[a] < hash[b]) ? +1 : (hash[a] > hash[b]) ? -1 : (b - a)
    })

    this.rankBit = value[0]<<16|value[1]<<12|value[2]<<8|value[3]<<4|value[4]
    this.hash = hash
    this.value = value
  }
}

module.exports = Hand

File: java/holdem/src/holdem/Hand.java
package holdem;

import java.util.*;
/**
 *
 * @author nilotpal
 */
public class Hand {
    /**
     * Stores dealt cards for a hand
     */
    String[] cards;
    
    private HashMap<String, Integer> hash = new HashMap<String, Integer>();
    private String suits = "";
    private Integer[] cardValues = new Integer[5];
    private String rank = "High Card";
    private int rankBit = 0;
    
    public Hand(String[] cards) {
        this.cards = cards;
        
        this.reduce();
        
        this.analyze();
    }
    
    public HashMap<String, Integer> hash () {
        return hash;
    }
    
    public Integer[] values () {
        return this.cardValues;
    }
    
    public String suits () {
        return this.suits;
    }
    
    public void setRank (String rank) {
        this.rank = rank;
    }
    
    public String rank () {
        return this.rank;
    }
    
    public int rankBit () {
        return this.rankBit;
    }
    
    public void analyze() {
        Rank analyzer = new Rank();
        analyzer.analyze(this);
    }
    
    private String[] toValue (String[] cards) {
        return Arrays.toString(cards).replaceAll("[CDHS ]", "").replace("[","").replace("]", "").split(",");
    }
    
    private int faceValue (String face) {
        return face.equals("T") ? 10 
                : face.equals("J") ? 11 
                : face.equals("Q") ? 12 
                : face.equals("K") ? 13 
                : face.equals("A") ? 14 
                : Integer.parseInt(face);
    }
    
    private void reduce () {
        //Extract card suits
        this.suits = Arrays.toString(this.cards).replaceAll("[^CDHS]", "");
        
        String[] cards = this.toValue(this.cards);

        for(int i=0; i < cards.length; i++) {
            cardValues[i] = faceValue(cards[i]);
        }
        
        for(int i=0; i < 5; i++) {
            String key = Integer.toString(cardValues[i]);
            hash.put(
                    key, 
                    (hash.get(key) != null && hash.get(key) >= 1) ? 
                            hash.get(key) + 1 : 1);
        }
        
        Arrays.sort(cardValues, new Comparator<Integer>() {
            @Override
            public int compare(Integer a, Integer b) {
                String aKey = Integer.toString(a);
                String bKey = Integer.toString(b);
                return (hash.get(aKey) < hash.get(bKey)) ? +1 : (hash.get(aKey) > hash.get(bKey)) ? -1 : (b - a);
            }
        });
        
        rankBit = cardValues[0]<<16|cardValues[1]<<12|cardValues[2]<<8|cardValues[3]<<4|cardValues[4];
    }
}
